zipkin-server/src/main/java/zipkin/server/internal/ZipkinHttpCollector.java

```diff
@@ -140,7 +147,13 @@
 
   static void error(HttpServerExchange exchange, Throwable e) {
     String message = e.getMessage();
-    int code = message == null || message.startsWith("Cannot store") ? 500 : 400;
+    int code;
+    if (e instanceof FilterActivatedException) {
+      FilterActivatedException filterException = (FilterActivatedException)e;
+      code = filterException.getHttpCode();
```

Comment:
@adriancole I tried to find a built-in exception class that carries something like http code, but unfortunately there isn't one that I can bring in without big dependencies. 

These two are the most commonly used: 
1) http://hc.apache.org/httpclient-3.x/apidocs/org/apache/commons/httpclient/HttpException.html
2) https://docs.oracle.com/javase/7/docs/api/javax/xml/ws/http/HTTPException.html

I still think the most reasonable option is to the custom exception. Without this, a filter implementor cannot adjust the http code emitted if there is an error, which would make for a bad user experience.

Reply:
I take part of this back. The latter exception will suffice for the error code and is bolted in java. The only drawback is that it doesn't provide a way to set a custom message.